src/asl.ts

```diff
@@ -497,12 +528,63 @@ export class ASL {
     return this.stateNames.get(stmt)!;
   }
 
-  public execute(stmt: Stmt): States {
+  private rewriteNext(
+    state: State,
+    scope: string,
+    scopeNames: Set<string>
+  ): State {
```

Comment:
Need some documentation on why we need to re-write. Examples/overall architecture information?

Reply:
This interpreter is starting to become difficult to track. Are we sure we need to re-write states like this versus perform pre-processing on the tree such that it can be mapped directly to ASL?